janus/spec/user_spec.rb

```diff
@@ -75,6 +75,79 @@
 
     expect(payload['flags']).to eq(user.flags.join(';'))
   end
+
+  it 'adds guest permission if user has agreed to coc, not on project already' do
+    user = create(
+      :user, name: 'Portunus', email: 'portunus@two-faces.org'
+    )
+    gateway = create(
+      :project, project_name: 'gateway', project_name_full: 'Gateway',
+      resource: true, requires_agreement: true
+    )
+    cc = create(
+      :cc_agreement, user_email: 'portunus@two-faces.org', cc_text: 'blah blah blah',
+      agreed: true, project_name: 'gateway'
+    )
+
+    user.set_guest_permissions!
+    expect(Permission.count).to eq(1)
+
+    perm = Permission.first
+    expect(perm.project_name).to eq('gateway')
+    expect(perm.user).to eq(user)
+    expect(perm.role).to eq('guest')
+  end
+
+  it 'does nothing if user has agreed to coc but already has viewer+ permission' do
+    user = create(
+      :user, name: 'Portunus', email: 'portunus@two-faces.org'
+    )
+    gateway = create(
+      :project, project_name: 'gateway', project_name_full: 'Gateway',
+      resource: true, requires_agreement: true
+    )
+    permission = create(
+      :permission, project: gateway, user: user, role: 'viewer'
+    )
+    cc = create(
+      :cc_agreement, user_email: 'portunus@two-faces.org', cc_text: 'blah blah blah',
+      agreed: true, project_name: 'gateway'
+    )
+
+    user.set_guest_permissions!
+    expect(Permission.count).to eq(1)
+
+    perm = Permission.first
+    expect(perm.project_name).to eq('gateway')
+    expect(perm.user).to eq(user)
+    expect(perm.role).to eq('viewer')
+  end
+  
+  it 'removes guest permission if user most recently did not agree to coc' do
```

Comment:
nit: might be nice to also include a test that some higher-level permission is **not** removed when not agreed to the cc. For example, if the user later is granted `viewer` access, after they've already declined the cc?

Reply:
:+1: ^^

Reply:
sure, will add!